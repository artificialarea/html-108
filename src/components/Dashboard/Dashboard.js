import React from 'react';
import uuid from 'react-uuid';
import { Link } from 'react-router-dom';


export default function Dashboard (props) {

    return (
        <div className="component dashboard">
            <DashboardHeader 
                who={props.who}
                userId={props.userId}
                users={props.users}
            />
            <TrackList 
                id={uuid()}  // TEMP until set up Routes
                who={props.who}
                userId={props.userId}
                users={props.users}
                tracks={props.tracks}
            />
        </div>
    )
}

function DashboardHeader (props) {

    let displayView;
    let usernameTitle;

    if (props.who !== 'private') {
        displayView = 'Community Dashboard'
    } else {
        usernameTitle = props.users.find(user => user.id === props.userId).username;
        displayView = `${usernameTitle}'s Dashboard`;
    }

    return (
        <header role="banner">
            <h1>{displayView}</h1>
            {props.who !== 'public' ? <Link to='/profile'>Edit Profile</Link> : null}
        </header>
    )
}

function TrackList (props) {

    const publicTracks = [];
    const privateTracks = [];

    props.tracks.forEach(track => {
        if (props.who !== 'private') {
            if (track.public === true) {
                const trackUser = props.users.find(user => user.id === track.user_id ).username
                publicTracks.push(
                    <TrackItem 
                        id={uuid()}
                        who={props.who}
                        user={trackUser} 
                        track={track}
                    />
                );
            }
        }
        if (track.user_id === props.userId) {
            privateTracks.push(
                <TrackItem 
                    id={uuid()}
                    who={props.who}
                    track={track}
                />
            );
        }
    })
    
    const dashboardDisplay = 
        (props.who !== 'private')
                ? publicTracks
                : privateTracks

    return (
        <ul className="track-list" key={props.id}>
            {dashboardDisplay}
        </ul>
    )
}

function TrackItem (props) {

    return (
        <li className="composition track-item" key={props.id}>
            <h2>{props.track.title}</h2>
            { props.who !== 'private' 
                ? <p className="public-user">by {props.user}</p> 
                : null }
            <p>3 Sept 2020</p>
            
            { props.who === 'private' && 
                <UserControls />
            }
            
            <button>Listen to Audio Sample</button> <button>See Drum Machine Configuration</button>
        </li>

    )
}

function UserControls (props) {
    return (
        <div className="user-controls">
            <input type="radio" name="composition-1" value="private"/>Private
            <input type="radio" name="composition-1" value="public"/>Public
            <br />
            <button>Delete Track</button>
        </div>
    )
}