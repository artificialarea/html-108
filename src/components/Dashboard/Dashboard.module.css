.root {
    background-color: #1e1e21;
    color: white;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
}
