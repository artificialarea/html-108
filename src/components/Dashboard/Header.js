import React from 'react';
import { Link } from 'react-router-dom';

export default function Header (props) {
    const { who, users, userId } = props;

    let displayView;
    let user;

    if (who !== 'private') {
        displayView = 'Community Dashboard'
    } else {
        user = users.find(user => user.id === userId).username;
        displayView = `${user}'s Dashboard`;
    }

    return (
        <header role="banner">
            <h1>{displayView}</h1>
            {who !== 'public' ? <Link to='/profile'>Edit Profile</Link> : null}
        </header>
    )
}