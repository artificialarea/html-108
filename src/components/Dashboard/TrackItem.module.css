.root {
    display: block;
    padding: 1em;
    background: rgba(40, 40, 44, 0.4);
    border-radius: 4px;
    margin: 10px;
    box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.3);
}

.root:hover {
    background: #28282c8c;
    background: rgba(40, 40, 44, 1)
}

.link {
    color: white;
    text-decoration: none;
}

.date {
    color: gray;
}