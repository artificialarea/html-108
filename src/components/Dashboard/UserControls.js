import React from 'react';
// import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
// import { 
//     faRedoAlt, 
//     faTrashAlt, 
//     faCloudDownloadAlt, 
//   } from '@fortawesome/free-solid-svg-icons';

export default function UserControls (props) {
    const { track } = props;

    return (
        <div className="user-controls">
            <label>
                <input 
                    type="radio" 
                    // name={`track-${track.id}`} 
                    name={track.id} 
                    value="private"
                    checked={!track.visible}
                    onChange={e => props.onChange(e)}

                />
                Private
            </label>
            <label>
                <input 
                    type="radio" 
                    name={track.id}  
                    value="public"
                    checked={track.visible}
                    onChange={e => props.onChange(e)}
                />
                Public
            </label>
            <br />
            <button
                type="button"
                onClick={() => props.onClickDelete(track.id)}
            >
                Delete Track
            </button>
            {/* <FontAwesomeIcon 
                icon={faTrashAlt}
                type="button"
                onClick={() => props.onClickDelete(track.id)}
            /> */}
        </div>
    )
}