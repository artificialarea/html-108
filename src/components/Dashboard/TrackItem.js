import React from 'react';
import { Link } from 'react-router-dom';
import UserControls from './UserControls';
import styles from './TrackItem.module.css';

export default function TrackItem (props) {
    const { track, user, who } = props;

    return (
        // <Link to={`/tracks/${track.id}`} className={styles.link}>    // bypassing ViewTrack view
        <Link to={`/edit/${track.id}`} className={styles.link}>
        <li className={styles.root}>
            <h2>{track.title}</h2>
            {/* { who !== 'private' 
                ? <p className="public-user">by {user}</p> 
                : null } */}
            <p><span className={styles.date}>{track.date_modified}</span></p>
            
            { who === 'private' && 
                <UserControls 
                    track={track}
                    onChange={e => props.onChange(e)}
                    // onClickDelete={props.onClickDelete}
                    onClickDelete={(trackId) => props.onClickDelete(trackId)}
                />
            }
            
            {/* <Link to={`/tracks/${track.id}`}>See Drum Machine Configuration</Link> */}

            {/* <button>Listen to Audio Sample</button> // revisit once dealing with sound */}
            
        </li>
        </Link>
    )
}