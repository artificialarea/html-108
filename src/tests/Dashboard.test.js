import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
// TODO: do some proper event listener snapshot tests with enzyme
// import { shallow } from 'enzyme';
// import toJson from 'enzyme-to-json';
import Dashboard from '../components/Dashboard/Dashboard';
import { BrowserRouter } from 'react-router-dom';

describe(`Dashboard component`, () => {

    // stub items for testing
    const usersProps = [
        {
            id: 1,
            username: 'admin_node',
            password: 'aaAA11!!',
            email: 'somebody@somewhere.com',
            date_created: '2029-01-22T16:28:32.615Z',
        },
    ];
    const tracksProps = [
        {  
            id: 1, 
            user_id: 1,
            title: 'Silencio', 
            date_modified: '2029-01-22T16:28:32.615Z',
            visible: true, 
            tempo: 120,
            sequence_length: 8,
            notes: ["G5", "Eb5", "C5", "G4"],
            checked: [
                [false,false,false,false,false,false,false,false],
                [false,false,false,false,false,false,false,false],
                [false,false,false,false,false,false,false,false],
                [false,false,false,false,false,false,false,false],
            ],
        },
        {  
            id: 2, 
            user_id: 1,
            title: 'FullBlastoff', 
            date_modified: '2100-05-22T16:28:32.615Z',
            visible: true,
            tempo: 200,
            sequence_length: 8,
            notes: ["G5", "Eb5", "C5", "G4"],
            checked: [
                [true,true,true,true,true,true,true,true],
                [true,true,true,true,true,true,true,true],
                [true,true,true,true,true,true,true,true],
                [true,true,true,true,true,true,true,true],
            ], 
        },
        {  
            id: 3, 
            user_id: 1,
            title: 'Browser Noise', 
            date_modified: '1919-12-22T16:28:32.615Z',
            visible: true,
            tempo: 80,
            sequence_length: 8,
            notes: ["G5", "Eb5", "C5", "G4"],
            checked: [
                [false,false,false,false,false,true,true,false],
                [true,false,true,false,false,false,false,true],
                [false,true,false,false,false,true,false,false],
                [true,false,true,false,true,false,true,false],
            ],
        },
    ];

    it('renders without crashing', () => {
        const div = document.createElement("div");
        ReactDOM.render(
            <BrowserRouter>
                <Dashboard />
            </BrowserRouter>
            , div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it(`renders the UI as expected`, () => {
        const tree = renderer
            .create(<BrowserRouter>
                <Dashboard 
                    who={'public'}
                    users={usersProps}
                    tracks={tracksProps}
                />
            </BrowserRouter>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

