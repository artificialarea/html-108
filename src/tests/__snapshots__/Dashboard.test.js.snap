// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component renders the UI as expected 1`] = `
<div
  className="root"
>
  <header
    role="banner"
  >
    <h1>
      Community Dashboard
    </h1>
  </header>
  <ul
    className="root"
  >
    <a
      className="link"
      href="/edit/1"
      onClick={[Function]}
    >
      <li
        className="root"
      >
        <h2>
          Silencio
        </h2>
        <p>
          <span
            className="date"
          >
            2029-01-22T16:28:32.615Z
          </span>
        </p>
      </li>
    </a>
    <a
      className="link"
      href="/edit/2"
      onClick={[Function]}
    >
      <li
        className="root"
      >
        <h2>
          FullBlastoff
        </h2>
        <p>
          <span
            className="date"
          >
            2100-05-22T16:28:32.615Z
          </span>
        </p>
      </li>
    </a>
    <a
      className="link"
      href="/edit/3"
      onClick={[Function]}
    >
      <li
        className="root"
      >
        <h2>
          Browser Noise
        </h2>
        <p>
          <span
            className="date"
          >
            1919-12-22T16:28:32.615Z
          </span>
        </p>
      </li>
    </a>
  </ul>
</div>
`;
